Задание 1
Импортируйте библиотеки pandas и numpy.
Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. Создайте датафреймы X и y из этих данных.
Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки
составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.
Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.
Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.
Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.


In [61]:
import pandas as pd
import numpy as np

In [62]:
from sklearn.datasets import load_boston

In [63]:
boston = load_boston()
data = boston["data"]

In [64]:
feature_names = boston["feature_names"]

X = pd.DataFrame(data, columns=feature_names)
X.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [65]:
target = boston["target"]

y = pd.DataFrame(target, columns=["price"])
y.head()

price
0   24.0
1   21.6
2   34.7
3   33.4
4   36.2

In [66]:
from sklearn.model_selection import train_test_split

In [67]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [68]:
from sklearn.linear_model import LinearRegression

In [69]:
lr = LinearRegression(copy_X=True)

In [70]:
lr.fit(X_train, y_train)

LinearRegression()

In [71]:
y_pred_lr = lr.predict(X_test)

In [72]:
from sklearn.metrics import r2_score

m1 = r2_score(y_true=Y_test, y_pred=y_pred_lr)

m1

0.7112260057484943

Задание 2
Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble.
Сделайте агрумент n_estimators равным 1000,
max_depth должен быть равен 12 и random_state сделайте равным 42.
Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression,
но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0],
чтобы получить из датафрейма одномерный массив Numpy,
так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.
Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания.
Напишите в комментариях к коду, какая модель в данном случае работает лучше.


In [73]:
from sklearn.ensemble import RandomForestRegressor

In [74]:
 model = RandomForestRegressor(
    n_estimators=1000,
    max_depth=12,
    random_state=42
)

In [75]:
X, y = load_boston(return_X_y=True)
X = pd.DataFrame(data=X)
y = pd.DataFrame(data=y)

In [76]:
 X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [77]:
model.fit(X_train, y_train.values[:, 0])

RandomForestRegressor(max_depth=12, n_estimators=1000, random_state=42)

In [78]:
y_pred = model.predict(X_test)
y_pred[:2], y_test.head(2)

(array([22.80641237, 31.13146352]),
         0
 173  23.6
 274  32.4)

In [79]:
m2 = r2_score(y_true=y_test, y_pred=y_pred.reshape(-1, 1))

m2

0.87472606157312

In [80]:
r2_scale = round(100 * (m2 - m1) / m2, 2)

r2_scale

18.69

Модель RandomForestRegressor дала более достоверный прогноз, чем LinearRegression на 18.69 %